jplag.frontend.rust/src/main/java/de/jplag/rust/JplagRustListener.java

```diff
@@ -0,0 +1,898 @@
+            case "else" -> {
+                if (stateContext == RustContext.IF_BODY) {
+                    transformToken(ELSE_STATEMENT, token);
+                    state.enter(RustContext.ELSE_BODY);
+                }
+            }
+            default -> {
+                // do nothing
+            }
+        }
+    }
+
+    @Override
+    public void enterType_(RustParser.Type_Context context) {
+        if (context.parent instanceof RustParser.GenericArgsTypesContext) {
+            transformToken(TYPE_ARGUMENT, context.getStart());
+        }
+
+        state.enter(RustContext.TYPE);
+        super.enterType_(context);
+    }
+
+    @Override
+    public void exitType_(RustParser.Type_Context context) {
+        state.leave(RustContext.TYPE);
+        super.exitType_(context);
+    }
+
+    @Override
+    public void enterGenericArg(RustParser.GenericArgContext context) {
+        // Only type arguments for methods, not for type expressions
+        if (context.getParent().getParent() instanceof RustParser.PathInExpressionContext) {
+            transformToken(TYPE_ARGUMENT, context.getStart());
+        }
+        super.enterGenericArg(context);
+    }
+
+    @Override
+    public void enterEveryRule(ParserRuleContext context) {
+        // ExpressionContext gets no own enter/exit method
+        // used in various 'lists' of elements
+        if (context instanceof RustParser.ExpressionContext expression) {
+            if (context.parent instanceof RustParser.ArrayElementsContext) {
+                transformToken(ARRAY_ELEMENT, expression.getStart());
+            } else if (context.parent instanceof RustParser.CallParamsContext) {
+                transformToken(ARGUMENT, expression.getStart());
+            } else if (context.parent instanceof RustParser.TuplePatternItemsContext || context.parent instanceof RustParser.TupleElementsContext) {
+                if (state.getCurrent() == RustContext.REDUNDANT_TUPLE)
+                    return;
+                transformToken(TUPLE_ELEMENT, expression.getStart());
+            } else if (context.parent instanceof RustParser.ClosureExpressionContext) {
+                transformToken(CLOSURE_BODY_START, context.getStart());
+                transformToken(RETURN, expression.getStart());
+            }
+        }
+    }
+
+    @Override
+    public void exitEveryRule(ParserRuleContext context) {
+        if (context instanceof RustParser.ExpressionContext && context.parent instanceof RustParser.ClosureExpressionContext) {
+            transformToken(CLOSURE_BODY_END, context.getStop());
+
+        }
+    }
+
+    /**
+     * Implementation of Context for the Rust language
+     */
+    enum RustContext implements ParserState.Context {
+        /**
+         * This is used to make sure that the stack is not empty -> getCurrent() != null
+         **/
+        FILE(NONE, NONE),
+
+        /**
+         * These contexts are used to assign the correct tokens to '{' and '}' terminals.
```

Comment:
Is that not possible via the exit methods?

Reply:
In some cases, yes. I'll make an example:
```rust
if (true) { println!("{}", 1); }
```
This statement is handled as follows:
1. Assign `IF_STATEMENT` in the `enter` method of the if statement
2. Traverse the condition and assign tokens as needed
3. Now, the `IF_BODY_START` token shall be added at the position of `{`. This can easily be done when the terminal `{` itself is traversed; another way would be to try to find its position in the `enter` method, but then the order of the tokens might turn out wrong.
4. Traverse the body and assign tokens as needed
5. Add the `IF_BODY_END` token. This could be done in the `exit` method of the if statement indeed.

In Rust, there are many instances of optional bodies (for `struct`s, for example) and cases where any type of bracket is permitted (e.g. macro calls: `()`, `[]`, `{}`), so it seemed easier to wait for the terminal to actually occur instead of trying to find out whether/where/which one occurs in the `enter` method.

Reply:
So there is only one enter and exit method for bodies, but not a specific one for if/for/... bodies?

Reply:
I was beginning to wonder, and looking at the other frontends again, I now remember why this sort of thing was easy before and is now hard.
There are no enter/exit methods for any bodies at all, because the bodies get no syntactical category of their own in the Rust grammar.

```antlr
// in C++ frontend
ifStatement: 
	IF OPEN_PARENS expression CLOSE_PARENS if_body (ELSE if_body)? ;
if_body: 
	block ;

// in Rust frontend
ifExpression:
	'if' expression blockExpression ( 'else' (blockExpression | ifExpression | ifLetExpression) )? ;
```
In the C++ grammar, the if-body gets a grammatical rule of its own, so that it can be easily distinguished from any other type of block.

In the Rust grammar however, the if-body is simply a blockExpression, but a blockExpression can occur in many different contexts. So, the grammar will not help to tell blockExpressions apart.

So, it is really the general formulation of the Rust grammar that does not play in my favor here. Your reply describes the problem perfectly.

Reply:
Ah, now that makes sense! Thank you for your insights. We may add a sentence or two regarding that in the class JDoc or the frontend readme.

Reply:
Probably out of scope for this PR, but if this is the entire problem, refactoring the grammar could also be an option. Based on your experience, what would you estimate how difficult this is? I would assume it to not be too hard as it should only introduce an intermediate rule. Or do I oversee something there?

Reply:
Yes, that should be entirely possible. I did that for another frontend but did not think to do it this time.